kotlin/workflow-ui-core/src/main/java/com/squareup/workflow/ui/BackStackScreen.kt

```diff
@@ -16,17 +16,59 @@
 package com.squareup.workflow.ui
 
 /**
- * @param stack: screens that are being or have been displayed, ending in the current screen.
- * Guaranteed not to be empty.
+ * Represents an active screen ([top]), and a set of previously visited
+ * screens to which we may return ([backStack]). By rendering the entire
+ * history we allow the UI to do things like maintain cached view state,
+ * implement drag-back gestures without waiting for the workflow, etc.
+ *
+ * Effectively a list that can never be empty.
  */
-data class BackStackScreen<StackedT : Any>(val stack: List<StackedT>) {
-  constructor(only: StackedT) : this(listOf(only))
-  constructor(vararg stack: StackedT) : this(stack.toList())
+class BackStackScreen<StackedT : Any>(
+  bottom: StackedT,
+  rest: List<StackedT>
+) {
+  constructor(
+    bottom: StackedT,
+    vararg rest: StackedT
+  ) : this(bottom, rest.toList())
+
+  val frames: List<StackedT> = listOf(bottom) + rest
+
+  /**
+   * The active screen.
+   */
+  val top: StackedT = frames.last()
+
+  /**
+   * Screens to which we may return.
+   */
+  val backStack: List<StackedT> = frames.subList(0, frames.size - 1)
```

Comment:
You can also do `frames.dropLast(1)`.

Reply:
Nice, thanks.

Reply:
Oh, wait a minute: those drop methods make new lists rather than just views on the original. Wasteful.

Reply:
Good point!